### PR TITLE
[supervisor] set pod failure reason when supervisor is reaped

### PR DESCRIPTION
**Pull request by mustard-mh**
## Description

- [x] Revert debug commit https://github.com/gitpod-io/gitpod/pull/20318/commits/1b6bde80c410034b62e235438b8f7262fa7f63b6 /hold

Root problem of CLC-877:

There can be a racing between `reaper.Reap()` and `runCommand.Wait()`. Our logic can process `runCommand.Wait()` well but once it's reaped by `reaper.Reap()` we know nothing (like exitCode). And the pod will exit without message or proper exitCode and then reach this line to update pod failed reason 

https://github.com/gitpod-io/gitpod/blob/1267b7c74bfd4cd92d84a91761ccaaab06076ce5/components/ws-manager-mk2/controllers/status.go#L386

We created a fork https://github.com/gitpod-io/go-reaper/pull/1, temporary add a function ourselves, and after upstream https://github.com/ramr/go-reaper/tree/notifier to release, we need a follow-up PR

## Related Issue(s)

Fixes CLC-877

## How to test

> [!NOTE]  
> Please use https://github.com/gitpod-io/empty to open workspaces to save space usage of preview env

- With debug commit applied https://github.com/gitpod-io/gitpod/pull/20318/commits/1b6bde80c410034b62e235438b8f7262fa7f63b6
- Start X workspaces with any JetBrains IDEs, and wait 2 more minutes. The workspace failed reason should be always `xxx timed out to start after 2 minutes` (as we don't know when the racing will happen, but it's very frequently with this case actually, ten times should be fine -> you could start 4 workspaces at the same time)
- Regular workspace stop should work like before
   - Exec`gp timeout set 1m`, to see it doesn't break timed out case
   - Exec `gp stop`
   - Make a special image and let it failed. i.e. https://github.com/kylos101/gitpod-custom-image/tree/kylos101/build-fail
   - ...

#### Preview status

<p>Gitpod was successfully deployed to your preview environment.</p>
<ul>
	<li><b>🏷️ Name</b> - hw-clc-877</li>
	<li><b>🔗 URL</b> - <a href="https://hw-clc-877.preview.gitpod-dev.com/workspaces" target="_blank">hw-clc-877.preview.gitpod-dev.com/workspaces</a>.</li>
	<li><b>📚 Documentation</b> - See our <a href="https://www.notion.so/gitpod/6debd359591b43688b52f76329d04010#7c1ce80ab31a41e29eff2735e38eec39" target="_blank">internal documentation</a> for information on how to interact with your preview environment.</li>
	<li><b>📦 Version</b> - hw-CLC-877-gha.29691</li>
	<li><b>🗒️ Logs</b> - <a href="https://console.cloud.google.com/logs/query;query=jsonPayload.kubernetes.host%3D%22preview-hw-clc-877%22%0A%0A--%20Filter%20on%20service:%0A--%20jsonPayload.serviceContext.service%3D%22ws-manager-mk2%22%0A;duration=P1D?project=gitpod-dev-preview" target="_blank">GCP Logs Explorer</a></li>
</ul>

## Build Options

<details>
<summary>Build</summary>

- [ ] /werft with-werft
      Run the build with werft instead of GHA
- [ ] leeway-no-cache
- [ ] /werft no-test
      Run Leeway with `--dont-test`
</details>

<details>
<summary>Publish</summary>

- [ ] /werft publish-to-npm
- [ ] /werft publish-to-jb-marketplace
</details>

<details>
<summary>Installer</summary>

- [ ] analytics=segment
- [ ] with-dedicated-emulation
- [ ] workspace-feature-flags
  Add desired feature flags to the end of the line above, space separated
</details>

<details>
<summary>Preview Environment / Integration Tests</summary>

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [ ] /werft with-large-vm
- [x] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [x] /werft preemptible
      Saves cost. Untick this only if you're really sure you need a non-preemtible machine.
- [ ] with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`. If enabled, `with-preview` and `with-large-vm` will be enabled.
- [ ] with-monitoring
</details>

/hold

### CODE REVIEW

**Review by filiptronicek — APPROVED** [review]
Code looks good, a brief suite of tests also didn't turn up anything unexpected. Left one relevant comment.

Nice work!

File: components/supervisor/cmd/init.go [hunk 1]
```diff
@@ -77,25 +77,55 @@ var initCmd = &cobra.Command{
...
-			if err != nil && !(strings.Contains(err.Error(), "signal: ") || strings.Contains(err.Error(), "no child processes")) {
+			if err == nil {
+				return
+			}
+			// exited by reaper
+			if strings.Contains(err.Error(), "no child processes") {
+				ctx, cancel := context.WithTimeout(context.Background(), time.Second*5)
+				defer cancel()
+				select {
+				case <-ctx.Done(): // timeout
```
[the remark this answers is not in the document]
Comment (mustard-mh) [reply components/supervisor/cmd/init.go:104]: > Should we have some special exit condition here?

There're reasons process exits: OS Signal, pod force kill, reaper.Reap(), common cases (runCommand exits)

We can't sure if we could receive `reaper.Reap()` 's event in time or other cases make it stop, so maybe doing nothing and fallback to old logic is the best choice: exit 0 or same old `containers of a workspace pod are not supposed to do that`

[the remark this answers is not in the document]
Comment (filiptronicek) [reply components/supervisor/cmd/init.go:104]: Sounds good since this is for sure an improvement. Let's keep as-is

**Comment by mustard-mh** [comment]
@filiptronicek Thanks for your review, I will revert debug commit later today and 🚢

File: components/supervisor/cmd/init.go [hunk 2]
```diff
@@ -77,25 +77,55 @@ var initCmd = &cobra.Command{
...
+					handleSupervisorExit(eerr.ExitCode())
 				}
 				log.WithError(err).Error("supervisor run error")
 				return
 			}
 		}()
 		// start the reaper to clean up zombie processes
-		reaper.Reap()
+		reaper.Start(reaper.Config{
+			Pid:              -1,
```
Comment (geropl) [inline components/supervisor/cmd/init.go:118]: What does this config mean? Would it make sense to add a comment for that?

**Review by geropl — APPROVED** [review]
Awesome find @mustard-mh ! :rocket: :sparkles: 

Test worked 10/10 :heavy_check_mark:

File: components/supervisor/cmd/init.go (hunk 2 above)
Comment (mustard-mh) [reply components/supervisor/cmd/init.go:118]: This config comes from default `reaper.Reap()` https://github.com/ramr/go-reaper/blob/d197ec784a27195ab4144ba839f450af47909776/reaper.go#L89-L98

**Comment by mustard-mh** [comment]
Debug commit reverted, smoke tested it can start JetBrains IDE after reverting. /unhold

image

**Pull request merged**